test/swarming/collect.py

```diff
@@ -0,0 +1,144 @@
+    assert os.path.exists(summary)
+
+    #### Extract test results
+    with open(summary, 'r') as f:
+        result = json.load(f)
+    # We expect a single shard
+    assert len(result['shards']) == 1
+    task_result = result['shards'][0]
+    device = ''
+    task_id = task_result['task_id']
+    # Find device name in the "tags" list, the entry looks like
+    # "device_type:<device>". The device is also mentioned the bot_dimensions
+    # list with a "device_type" key, but the value of this dimension is itself a
+    # list for which we have no guarantee on the order. Hence we prefer to
+    # retrieve the device name from the tags.
+    for tag in task_result['tags']:
+        if tag.startswith('device_type:'):
+            device = tag[len('device_type:'):]
+    assert device != ''
+    # Set status: default to fail, and lower the risk of false-positive by being
+    # pedantic to setup the 'pass' status
+    status = 'fail'
+    if (task_result['exit_code'] == '0') and (task_result['state'] == 'COMPLETED') and (task_result['failure'] == False) and (task_result['internal_failure'] == False):
+        status = 'pass'
+    elif task_result['state'] == 'TIMEOUT':
+        status = 'timeout'
+    elif task_result['state'] == 'EXPIRED':
+        status = 'expired'
+    elif task_result['internal_failure']:
+        status = 'internal_failure'
+
+    #### Add result to result file
+    # Results are stored in JSON with the following format:
+    # {
+    #   "20200324-032345": { # Build timestamp, YYYYMMDD-HHMMSS to be usable as a key to sort
+    #     "commit": "a1b2c3e4f5",  # build commit sha
+    #     "tests": {
+    #       "flame": { # device name
+    #         "foo-bar": { # test name
+    #           "task_id": "abcd1234", # Swarming task ID
+    #           "status": "pass"      # Swarming task status (pass/fail/expired...)
+    #         },
+    #         ...
+    #       }
+    #     }
+    #   }
+    # }
+
+    if os.path.exists(args.results_json):
+        with open(args.results_json, 'r') as f:
+            results = json.load(f)
+    else:
+        print('Warning: results file "' + args.results_json + '" does not exists, it will be created')
+        results = {}
+
+    if args.build_timestamp in results.keys():
+        assert results[args.build_timestamp]['commit'] == args.commit
+    else:
+        results[args.build_timestamp] = {
+            'commit': args.commit,
+            'tests': {}
+        }
+
+    if device not in results[args.build_timestamp]['tests'].keys():
+        results[args.build_timestamp]['tests'][device] = {}
+    assert args.test_name not in results[args.build_timestamp]['tests'][device].keys()
+    results[args.build_timestamp]['tests'][device][args.test_name] = {
+        'task_id': task_id,
+        'status': status
+    }
+
+    #### Write back result file
+    os.makedirs(os.path.dirname(args.results_json), exist_ok=True)
+    with open(args.results_json, 'w') as f:
```

Comment:
You want to append to that file when the file existed, right? Wouldn't it be better to open it with "append"?
```
open(args.results_json, "a")...
```

Reply:
Ah sorry, the `results` variable already contains the previous results.